### PR TITLE
Feature/multi sort

### PR DESCRIPTION
**Pull request by Diiegoav**
## Related Issue

#585 Multiple Sorting

## Description

 - Removed single sorting, to apply multi column sorting, based on `maxColumnSortProp`.
 - Added new `initialOrderCollection` to specify an initial multi column sort.
 - Added new `showColumnSortOrder & sortOrderIndicatorStyle`, for show the order indicator of clicked column, and be able to styled it need it.
 - Allow the user to still use `columns.options.customSort`.
 - Kept the current `columns.option.sorting` flag to turn on/off any column from sorting.
 - Deprecated `options.sorting`, `query.orderBy, query.orderDirection` and `onOrderChange`.

Works on normal, tree and group table, for group table groups child data only supports single sort.

## Related PRs

None

## Impacted Areas in Application

List general components of the application that this PR will affect:

\* src/material-table.js
\* src/utils/data-manager.js
\* src/components/MTableHeader/index.js

## Additional Notes

Please review the pr, let me know your thoughts, and if there is any feedback let me know to update the pr, I think this multicolumn sort could be a great contribution for all the community that uses material table :)

### CODE REVIEW

File: src/utils/data-manager.js [hunk 1]
```diff
@@ -3,6 +3,8 @@ import uuid from 'uuid';
 import { selectFromObject } from './';
 import { widthToNumber } from './common-values';
 
+const MANY = 'many';
```
Comment (Domino987) [inline src/utils/data-manager.js]: Many? OR all?, and is this exposed somewehre for the user to import it?

File: src/utils/data-manager.js [hunk 2]
```diff
@@ -183,6 +185,31 @@ export default class DataManager {
     this.defaultExpanded = expanded;
   }
 
+  setMaxColumnSort(maxColumnSort) {
+    const notAvailableColumns = this.columns.filter(
+      (column) => column.sorting === false
+    );
+    const availableColumnsLength =
```
Comment (Domino987) [inline src/utils/data-manager.js]: const availableColumnsLength = this.columns.filter(
      (column) => column.sorting !== false
    ).length;

File: src/utils/data-manager.js [hunk 3]
```diff
@@ -183,6 +185,31 @@ export default class DataManager {
...
+      this.columns.length - notAvailableColumns.length;
+
+    if (maxColumnSort === MANY) {
+      this.maxColumnSort = availableColumnsLength.length;
+    } else {
+      this.maxColumnSort =
+        maxColumnSort > availableColumnsLength
+          ? availableColumnsLength
```
Comment (Domino987) [inline src/utils/data-manager.js]: this.maxColumnSort = Math.min(maxColumnSort, availableColumnsLength)

File: src/utils/data-manager.js [hunk 4]
```diff
@@ -369,11 +396,55 @@ export default class DataManager {
...
-    this.currentPage = 0;
+  getOrderByCollection = () => {
+    return this.orderByCollection.filter((collection) => collection.sortOrder);
+  };
+
+  sortOrderCollection = (list) => {
+    return list.sort((a, b) => {
+      if (!a.sortOrder) return 1;
+      if (!b.sortOrder) return -1;
+      return a.sortOrder < b.sortOrder ? -1 : a.sortOrder > b.sortOrder ? 1 : 0;
```
Comment (Domino987) [inline src/utils/data-manager.js]: return b.sortOrder - a.sortORder?

File: src/utils/data-manager.js [hunk 5]
```diff
@@ -369,11 +396,55 @@ export default class DataManager {
...
+  changeColumnOrder(orderBy, orderDirection, sortOrder) {
+    let prevColumns = [];
+    const sortColumns = this.getOrderByCollection();
+
+    if (sortColumns.length === this.maxColumnSort && !sortOrder) {
+      this.orderByCollection[0].orderDirection = '';
+      this.orderByCollection[0].sortOrder = undefined;
+
+      prevColumns = this.orderByCollection.map((collection) => {
+        if (collection.sortOrder) {
```
Comment (Domino987) [inline src/utils/data-manager.js:414]: sortOrder 0 will be false here too?

File: src/utils/data-manager.js [hunk 6]
```diff
@@ -707,40 +778,63 @@ export default class DataManager {
...
 
-  sortList(list) {
-    let columnDef = this.columns.find((_) => _.tableData.id === this.orderBy);
-    if (!columnDef) {
-      columnDef = this.columns[0];
-    }
-    let result = list;
+  sortBy(list, columnsDefs, orderCollection) {
+    const sort = this.sort;
+    const getFieldValue = this.getFieldValue;
```
Comment (Domino987) [inline src/utils/data-manager.js:788]: Why is this extracted?

File: src/utils/data-manager.js [hunk 7]
```diff
@@ -1188,9 +1282,9 @@ export default class DataManager {
             element.groupsIndex = getGroupsIndex(element.groups);
             sortGroupData(element.groups, level + 1);
           } else {
-            if (this.orderBy >= 0 && this.orderDirection) {
+            if (this.maxColumnSort > 0 && this.orderByCollection.length) {
               element.data = this.sortList(element.data);
-            } else if (this.orderDirection === '') {
+            } else if (!this.orderByCollection.length) {
```
Comment (Domino987) [inline src/utils/data-manager.js]: his.orderByCollection.length > 0 ?

File: src/utils/data-manager.js [hunk 8]
```diff
@@ -1209,7 +1303,7 @@ export default class DataManager {
       sortGroupData(this.sortedData, 1);
     } else if (this.isDataType('tree')) {
       this.sortedData = [...this.treefiedData];
-      if (this.orderBy != -1) {
+      if (this.maxColumnSort > 0 && this.orderByCollection.length) {
```
Comment (Domino987) [inline src/utils/data-manager.js]: Can we try not to have falsy comparisons of numbers?

**Review by Domino987 — CHANGES_REQUESTED** [review]

File: src/utils/data-manager.js (hunk 1 above)
Comment (Diiegoav) [reply src/utils/data-manager.js]: All could be a good fit, let me updated and expose it.

File: src/utils/data-manager.js (hunk 2 above)
Comment (Diiegoav) [reply src/utils/data-manager.js]: with this `column.sorting !== false` we will make users to only disabled sorting when type `sorting = false`, is that expected ?

File: src/utils/data-manager.js (hunk 5 above)
Comment (Diiegoav) [reply src/utils/data-manager.js:414]: yes, it will not enter to the is statement reduce its count.

File: src/utils/data-manager.js (hunk 6 above)
Comment (Diiegoav) [reply src/utils/data-manager.js:788]: I thought it was going to be reused, but I thinks it not, let me revert it back to the sort function.

**Review by Domino987 — APPROVED** [review]

**Pull request merged**